pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -1963,3 +1969,26 @@ func (l *LibvirtDomainManager) GetFilesystems() ([]v1.VirtualMachineInstanceFile
 
 	return fsList, nil
 }
+
+func detachHostDevices(virConn cli.Connection, dom cli.VirDomain) error {
+	domainSpec, err := util.GetDomainSpecWithFlags(dom, 0)
+	if err != nil {
+		return err
+	}
+
+	if registrar := cli.NewDomainEventDeviceRemovedRegistrar(virConn, dom); registrar != nil {
+		err := detachSRIOVDevices(domainSpec, registrar, dom)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func detachSRIOVDevices(domainSpec *api.DomainSpec, registrar *cli.DomainEventDeviceRemovedRegistrar, dom cli.VirDomain) error {
+	sriovDevices := sriov.FilterHostDevices(domainSpec)
+
+	return sriov.WaitHostDevicesToDetach(registrar, sriovDevices, 30*time.Second, func() error {
+		return sriov.DetachHostDevices(dom, sriovDevices)
+	})
```

Comment:
What do you think splitting `WaitHostDevicesToDetach` in two, I think it has too much responsabilities (register, detach, select, unregister), so it would be something like 

```golang
detachChannel, err := sriov.RegisterToDevicesDetach(...
...
err = sriov.DetachHostDevices(...)
...
sriov.WaitHostDevicesToDetach(detachChannel)
```

Also we can use the golang context Timeout and Cancel mechanis to implement timeout and also the deregister as defer.

Reply:
`WaitHostDevicesToDetach` is abstracting all the details of **how** it is done.
Looking at the caller, it just needs to know how to use the `WaitHostDevicesToDetach` for blocking on the operation (`sriov.WaitHostDevicesToDetach`) to succeed (or not). The details of registration, channeling, etc is an implementation detail.

If you have a suggestion on how to extract the implementation of `WaitHostDevicesToDetach` into multiple functions, then lets move the discussion there. I did not have a good idea how to do it there.

Regarding the context, I think it fits go-routine usages (the caller using it as such), but here this is a blocking operation that will return on timeout which is defined internally. I do not see the advantage of a context in this case, as we do not handle the timeout from outside.

Reply:
Added a proposal on the correct place you can close this comment.